keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+`ValidatingAdmissionPolicy`.
+
+Pros:
+
+- Policy author doesn't need to define the CRD. It can instead be generated for them.
+
+Cons:
+
+- This implies that the policy owns the configuration. But we have use cases
+  where multiple policies share the same configuration. For this model it seems
+  misleading and potentially to have both policies defining the configuration.
+
+##### Match Criteria
+
+During admission, the Kubernetes API server will validate the resource being
+admitted against all policy configuration resources that match the resource.
+
+Match criteria must be declared in the `spec.match` field of policy
+configuration resources (see `ReplicaLimit` in the above example).
+
+In order for policy configuration resources to declare match criteria, the
+corresponding configuration CRD schema must has a `spec.match` property. This
+property must conform to the below "matching schema template". This ensures that
+the match criteria is in the format that API server expects (the API server will
+be using duck typing here since there is no established way to do polymorphism
+across CRDs). The schemas of these fields in the configuration CRD may omit any
+optional properties; policy definition authors should only include the parts of
+the "match schema template" that are useful for configuring a particular policy.
+
+(Also, by allowing the "matching schema template" in configuration CRDs to be a
+omit optional properties, this API is future proofed against the addition of
+other match related properties in the future).
+
+"matching schema template":
+
+```go
+// TODO: Add this as a struct into the Kubernetes codebase so it can easily be
+// imported?
+type Match struct {
+  Rules []admissionregisterationv1.RuleWithOperations `json:"rules,omitempty" protobuf:"bytes,1,rep,name=rules"`
+  NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+  ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,3,opt,name=objectSelector"`
+}
+```
+
+Example usage:
+
+```go
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ReplicaLimit struct {
+  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`
+
+  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
+}
+```
+
+For comparison (provided by @liggitt):
+
+- RBAC, priority & fairness, and admission webhook configurations all use
+ordered lists of rules, and first match allows/matches.
+- Audit configuration also uses an ordered list of rules, but can make a match
+exclude auditing... this approach could address the "ability to skip specific
+resource types" mentioned above.
+
+TODO: What exact matching mechanisms will be supported? Right now this KEP
+shows the use of the same match rules that admission webhooks use.
+
+TODO: Should policy definitions also have match criteria? Match criteria at the
```

Comment:
As currently written, the KEP shows policy definitions with webhook-style "match criteria" (operation, resource, etc.)

It's worth considering whether the role of "policy match criteria" is purely for type information, or if it can be used more generally (e.g. prevent any policy from touching `kube-system`). This discussion overlaps heavily with the notion of "standalone policies" below.

Reply:
I've added a table that shows the capabilities of a wider range of match rules found in kubernetes (webhooks, rbac, audit, priority&fairness).  We do want to go beyond just type information.  It's clear to me that webhook match criteria doesn't do enough. Lack of exclude option, in particular, is problematic (it has made excluding kube-system difficult). Lack of resource name matching is also inconvenient.